app/controllers/tests_controller.rb

```diff
@@ -34,90 +34,112 @@ def update
   def create
     # create test and run validations
     @test = Test.create!(test_params)
+    determine_baseline_test(@test, test_params[:screenshot])
+    # force save so that dragonfly does it persistence on the baseline image
+    @test.save!
+    temp_paths = temp_screenshot_paths(@test)
+    compare_result = compare_images(@test, temp_paths)
+    @test.pass = determine_pass(@test, temp_paths, compare_result)
+    save_or_discard_screenshots(@test, temp_paths)
+    @test.save
+    remove_temp_files(temp_paths)
+
+    # TODO: why are we rescuing this? Can we fix the problem?
+    begin
+      @test.create_thumbnails
+    rescue
+    end
+
+    render json: @test.to_json
+  end
+
+  private
+
+  def test_params
+    params.require(:test).permit(:name, :platform, :browser, :size, :screenshot, :run_id, :source_url, :fuzz_level)
+  end
+
+  def convert_image_command(input_file, output_file, canvas)
+    "convert #{input_file.shellescape} -background white -extent #{canvas[:width]}x#{canvas[:height]} #{output_file.shellescape}"
+  end
+
+  def compare_images_command(baseline_file, compare_file, diff_file, fuzz, highlight_colour)
+    "compare -alpha Off -dissimilarity-threshold 1 -fuzz #{fuzz} -metric AE -highlight-color #{highlight_colour} #{baseline_file.shellescape} #{compare_file.shellescape} #{diff_file.shellescape}"
+  end
 
+  def determine_baseline_test(test, screenshot)
     # find an existing baseline screenshot for this test
-    baseline_test = Test.find_baseline_by_key(@test.key)
+    baseline_test = Test.find_baseline_by_key(test.key)
 
     if baseline_test
       # grab the existing baseline image and cache it against this test
-      @test.screenshot_baseline = baseline_test.screenshot
+      test.screenshot_baseline = baseline_test.screenshot
     else
       # otherwise if no baseline exists (i.e. this is the first run of this test), mark test as the baseline
-      @test.baseline = true
-      @test.screenshot_baseline = test_params[:screenshot]
+      test.baseline = true
+      test.screenshot_baseline = screenshot
     end
 
-    # force save so that dragonfly does it persistence on the baseline image
-    @test.save!
+  end
 
-    # get the width/height of both the baseline and the test image
-    baseline_screenshot_details = ImageGeometry.new(@test.screenshot_baseline.path)
-    test_screenshot_details = ImageGeometry.new(@test.screenshot.path)
+  def create_canvas(test)
     # create a canvas using the baseline's dimensions
-    canvas = Canvas.new(baseline_screenshot_details, test_screenshot_details)
-    @test.dimensions_changed = canvas.dimensions_differ
```

Comment:
This just flags whether we had to resize the canvas, but yeah not really necessary any more. Can go.